p4c emit Counter and DirectCounter externs in json dump

derek-so opened the pull request:
Implemented Counters and Direct counters in p4c.

derek-so committed:
added extern instances into json output, buggy, incomplete

derek-so committed:
counter extern formatting v2

derek-so committed:
Counters and Direct counters added to JSON extern_instances node

derek-so committed:
reverting misc changes

derek-so committed:
fixed all cpplint errors

peteli3 commented on backends/bmv2/psa_switch/psaSwitch.h:
```diff
@@ -47,7 +47,35 @@ class PsaSwitchExpressionConverter : public ExpressionConverter {
+          } else if (fieldName == "PACKETS") {
+            cstring repr = BMV2::stringRepr(1, ROUNDUP(bitwidth, 32));
+            jsn->emplace("value", repr);
+          } else {
```
maybe it'd be nice to do this as
```
} else if (fieldName == "PACKETS_AND_BYTES") {
    // do your stuff
} else { 
    // yell at programmer for invalid counter type
}
```

derek-so replied:
Yea that would be nice, but Im not sure if the compiler would have caught something like that a lot earlier on. Psa_CounterType_t is an enum in psa.p4, and trying to access a member not identified there might already throw an error in the frontend / type checking passes? Should definitely ask the p4 people when we PR them

derek-so replied:
Yup changes made

peteli3 reviewed (changes requested):
maybe it'd be nice to do this as
```
} else if (fieldName == "PACKETS_AND_BYTES") {
    // do your stuff
} else { 
    // yell at programmer for invalid counter type
}
```

peteli3 commented:
I recall antonin mentioning something about ensuring that we should also make sure to include some counter from behavioral-model/src/bm_sim so that we can enable using switch CLI for debugging...  but perhaps it'd be best if we clarify w/ him.

Otherwise, LGTM

liujed committed:
Clear the type map after parsing annotations. (#1759)

Fixes #1697.

jafingerhut committed:
Add option to STF "expect" command for stricter checking of output packets (#1763)

Before and after this change, the default behavior for the STF
"expect" command is to check that the actual output packet from BMv2
is at least as long as the one specified in the "expect" statement,
and all bytes in the "expect" packet data are present at the beginning
of the actual output packet.  If the actual output packet has
additional data after that, the test still passes.

With this change, if you add a "$" character at the end of the packet
data in an "expect" statement, the test will fail in all cases where
"expect" fails today, plus it will also fail if the actual output
packet is longer than the one specified in the "expect" statement.
This enables slightly more strict checking of actual output packets.

Commit:
Support for 'int' type (#1618)

* Support for 'int' type

Commit:
Support for header unions in bmv2 simple_switch back-end; multiple test examples (#1760)

* Support for union stacks; multiple test examples

derek-so committed:
faulty input checks

Commit:
Implement sizeInBits (#1751)

* Allow fields called isValid and setValid in headers
* Implement sizeInBits; fixes #1717
* Renamed to minSizeIn*; added to stacks

derek-so committed:
Merge pull request #3 from p4lang/master

LatestUpdates

derek-so committed:
pr2

derek-so committed:
Merge branch 'dts76-dev' of https://github.com/peteli3/p4c into dts76-dev

Mergining changes from p4c master to use new helpers

derek-so committed:
using minWidthBits

derek-so committed:
correct usage of minBitWidth
